wagtail/models/__init__.py

```diff
@@ -217,7 +217,171 @@ class Meta:
+        user=None,
+        submitted_for_moderation=False,
+        approved_go_live_at=None,
+        changed=True,
+        log_action=False,
+        previous_revision=None,
+        clean=True,
+    ):
+        """
+        Creates and saves a revision.
+        :param user: the user performing the action
+        :param submitted_for_moderation: indicates whether the object was submitted for moderation
+        :param approved_go_live_at: the date and time the revision is approved to go live
+        :param changed: indicates whether there were any content changes
+        :param log_action: flag for logging the action. Pass False to skip logging. Can be passed an action string.
+            Defaults to 'wagtail.edit' when no 'previous_revision' param is passed, otherwise 'wagtail.revert'
+        :param previous_revision: indicates a revision reversal. Should be set to the previous revision instance
+        :param clean: Set this to False to skip cleaning object content before saving this revision
+        :return: the newly created revision
+        """
+        if clean:
+            self.full_clean()
+
+        comments = getattr(self, COMMENTS_RELATION_NAME, None)
+        if comments:
+            new_comments = comments.filter(pk__isnull=True)
+            for comment in new_comments:
+                # We need to ensure comments have an id in the revision, so positions can be identified correctly
+                comment.save()
+
+        # Create revision
+        revision = Revision.objects.create(
+            content_type=self.get_content_type(),
+            base_content_type=self.base_content_type,
+            object_id=self.pk,
+            submitted_for_moderation=submitted_for_moderation,
+            user=user,
+            approved_go_live_at=approved_go_live_at,
+            content=self.serializable_data(),
+        )
+        # This is necessary to prevent fetching a fresh object instance when
+        # changing first_published_at in Page.
+        # Ref: https://github.com/wagtail/wagtail/pull/3498
+        revision.content_object = self
+
+        update_fields = []
+
+        if comments:
+            for comment in new_comments:
+                comment.revision_created = revision
+
+            update_fields = [COMMENTS_RELATION_NAME]
+
+        if hasattr(self, "latest_revision_created_at"):
+            self.latest_revision_created_at = revision.created_at
+            update_fields.append("latest_revision_created_at")
+
+        if hasattr(self, "draft_title"):
+            self.draft_title = self.title
+            update_fields.append("draft_title")
+
+        if hasattr(self, "has_unpublished_changes") and changed:
+            self.has_unpublished_changes = True
+            update_fields.append("has_unpublished_changes")
+
+        if update_fields:
+            # clean=False because the fields we're updating don't need validation
+            self.save(update_fields=update_fields, clean=False)
```

Comment:
These seem to be specific to `Page`.

Looking at how this is written, I wonder if it's worth adding something like `{before,after}_create_revision` hooks here, and let `Page` define this logic?